test/restore-object.js

```diff
@@ -0,0 +1,85 @@
+"use strict";
+/* eslint-disable no-empty-function */
+
+var referee = require("@sinonjs/referee");
+var sinonSpy = require("../lib/sinon/spy");
+var sinonStub = require("../lib/sinon/stub");
+var restoreObject = require("../lib/sinon/restore-object.js");
+var assert = referee.assert;
+var refute = referee.refute;
+
+describe("restore-object", function() {
+    it("is defined", function() {
+        assert.isFunction(restoreObject);
+    });
+
+    it("works on empty objects", function() {
```

Comment:
Why should it? Sinon usually throws a meaninful exception in cases where passed arguments are not "Sinon objects". This helps finding issues while writing or refactoring test cases. Think of passing a non-fake object to `sinon.assert.called(xyz)`, for example. In case of this method I would prefer having an exception thrown if nothing is being restored. Other thoughts?

Reply:
@mantoni This is the way both `spyOnEntireObject` and `stubEntireObject` are defined. Should those change as well?

```shell
➜  sinon git:(restore-object) ✗ ag -Q 'walk(object || {}' lib
lib/sinon/stub-entire-object.js
7:    walk(object || {}, function(prop, propOwner) {

lib/sinon/spy-entire-object.js
7:    walk(object || {}, function(prop, propOwner) {

lib/sinon/restore-object.js
7:    walk(object || {}, function(prop, propOwner) {
```

Reply:
I lean towards failing in the new API. We can make it consistent now, or later as a breaking change. I have no specific opinion here. Do @fatso83 and @mroderick have additional arguments?

Reply:
What does failing do/mean? 

From my perspective, the user asks to restore all mocks, spies, and stubs. We return an object with 0 mocks, spies, and stubs.

The error case makes sense for 1-to-1 mappings of "do X to this spy/stub". It feels a bit more ambiguous here though.

Reply:
Break things! Yeah! I always like to land on the strict side of things, so :+1: to not allowing stubbing or restoring `undefined`.

Reply:
But ... we could do it in a separate PR just to not bloat this one.

Reply:
I like the idea of doing them outside of this fix. 

We should make them all congruent. That sounds like a very breaking change.

Reply:
It makes sense to keep the implementations symetric for now add strictness later. I have created #2090 summarizing what was discussed here. 👍

Reply:
I started working on #2090 in a branch(based off of this one).